Userland/Utilities/sed.cpp

```diff
@@ -0,0 +1,559 @@
+    {
+        auto& builder = format_builder.builder();
+        builder.appendff("{}", command.addr1);
+        if (command.addr2.address_type() != AddressType::Unset) {
+            builder.appendff(",{}", command.addr2);
+        }
+        builder.append(command.function);
+        builder.append(command.arguments);
+    }
+};
+}
+
+struct AddressParsingResult {
+    Optional<Address> address;
+    size_t end_of_address;
+};
+
+static AddressParsingResult parse_address(StringView sv)
+{
+    if (sv.is_empty())
+        return { {}, 0 };
+
+    if (sv[0] == '$') {
+        return { Address { AddressType::LastLine }, 1 };
+    }
+
+    size_t index = 0;
+    for (; index < sv.length() && '0' <= sv[index] && sv[index] <= '9'; index++)
+        ;
+    if (index == 0)
+        return { {}, 0 };
+    return {
+        Address { AK::StringUtils::convert_to_uint<size_t>(sv.substring_view(0, index)).release_value() },
+        index
+    };
+}
+
+template<int max_addresses>
+static bool verify_number_of_addresses(const Command& command)
+{
+    if constexpr (max_addresses == 2)
+        return true;
+    else {
+        static_assert(max_addresses == 0 || max_addresses == 1);
+        auto c = command.function;
+        if constexpr (max_addresses == 0) {
+            if (command.addr1.address_type() != AddressType::Unset) {
+                Core::File::standard_error()->write(String::formatted("'{}' doesn't take any address\n", c));
+                return false;
+            }
+        } else {
+            if (command.addr2.address_type() != AddressType::Unset) {
+                Core::File::standard_error()->write(String::formatted("'{}' takes a single address\n", c));
+                return false;
+            }
+        }
+    }
+    return true;
+}
+
+struct CommandParsingResult {
+    Optional<Command> command;
+    size_t last_index;
+};
+
+static CommandParsingResult parse_command(StringView command_p)
+{
+    Command command;
+    StringView to_parse = command_p;
+    size_t end_of_command_index = 0;
+
+    auto [maybe_address, end_of_address] = parse_address(to_parse);
+    end_of_command_index += end_of_address;
+    command.addr1 = maybe_address.value_or({});
+    if (end_of_address >= to_parse.length())
+        return { {}, end_of_command_index };
+    to_parse = to_parse.substring_view(end_of_address);
+
+    if (to_parse.starts_with(',')) {
+        to_parse = to_parse.substring_view(1);
+        auto [maybe_address, end_of_address] = parse_address(to_parse);
+        end_of_command_index += end_of_address + 1;
+        to_parse = to_parse.substring_view(end_of_address);
+        command.addr2 = maybe_address.value_or({});
+    }
+
+    if (to_parse.is_empty()) {
+        return { command, end_of_command_index };
+    }
+
+    char command_char = to_parse[0];
+    to_parse = to_parse.substring_view(1);
+    end_of_command_index += 1;
+#define HANDLE_FUNCTION_CASE(c, max_addresses)                     \
```

Comment:
Let's leave a bit of whitespace above this, it looks a little too crowded here :P

Reply:
Added, it was unreadable indeed